doc/sphinx/proofs/automatic-tactics/auto.rst

```diff
@@ -424,50 +391,55 @@ automatically created.
          .. coqtop:: reset all
 
             Require Import List.
-            Hint Extern 5 ({?X1 = ?X2} + {?X1 <> ?X2}) => generalize  X1, X2; decide equality : eqdec.
+            Hint Extern 5 ({?X1 = ?X2} + {?X1 <> ?X2}) => generalize  X1, X2;
+              decide equality : eqdec.
             Goal forall a b:list (nat * nat), {a = b} + {a <> b}.
             Info 1 auto with eqdec.
 
-   .. cmdv:: Hint Cut @regexp : @ident
-      :name: Hint Cut
+   .. cmd:: Hint Cut [ @hints_regexp ] {? : {+ @ident } }
+
+      .. insertprodn hints_regexp hints_regexp
+
+      .. prodn::
+         hints_regexp ::= {+ @qualid }
+         | _
+         | @hints_regexp %| @hints_regexp
+         | @hints_regexp @hints_regexp
+         | @hints_regexp *
+         | emp
+         | eps
+         | ( @hints_regexp )
+
+      The `emp` regexp does not match any search path while `eps`
+      matches the empty path.
+
+      Used to cut the proof search tree according to a regular
+      expression that matches the paths to be cut. Note that there is no
+      operator precedence during parsing, one can
+      check with :cmd:`Print HintDb` to verify the current cut expression:
 
       .. warning::
 
          These hints currently only apply to typeclass proof search and the
          :tacn:`typeclasses eauto` tactic.
 
-      This command can be used to cut the proof-search tree according to a regular
-      expression matching paths to be cut. The grammar for regular expressions is
-      the following. Beware, there is no operator precedence during parsing, one can
-      check with :cmd:`Print HintDb` to verify the current cut expression:
+      .. todo PR: Should we keep/change 79 uses of nonbreaking space in the rsts?
 
-      .. prodn::
-         regexp ::= @ident   (hint or instance identifier)
-         | _   (any hint)
-         | @regexp | @regexp   (disjunction)
-         | @regexp @regexp   (sequence)
-         | @regexp *   (Kleene star)
-         | emp   (empty)
-         | eps   (epsilon)
-         | ( @regexp )
```

Comment:
By removing the annotations, a lot of information has been lost. Thus, it should be reintroduced, for instance as a table.

Reply:
As I think you know, insertprodn/prodn doesn't make it easy to have annotations here.  Here's a crude and effective approach; the other way to do it to edit every line in common.edit_mlg plus a small code change to doc_grammar.ml.

Reply:
How do you set the cut expression back to `emp`?  Would you set this for a single proof or for larger chunks?

Reply:
My suggestion was not to reintroduce the annotations in the `prodn` but rather to add a separate table that explains each production, like what is done in the documentation of `Tactic Notation`.

> How do you set the cut expression back to `emp`? Would you set this for a single proof or for larger chunks?

I don't think you can. You usually do this for every proof. I.e., you add hints / instances that would create trouble if they were used in any circumstances and then you add a `Hint Cut` to limit the way they can be used.

Reply:
Yeah.  I think we should show the productions and that a table in addition is overkill.  (You only asked that I reintroduce the information, while leaving the means up to me.)  Is that agreeable?

> You usually do this for every proof. I.e., you add hints / instances that would create trouble if they were used in any circumstances and then you add a `Hint Cut` to limit the way they can be used.

How do you make the following work for proof #2?  Would you make the `Hint Cut` local within a section just for proof #1?

```
Hint Cut foo.
proof #1
Hint Cut bar.  (now has the cut expression "foo | bar")
proof #2  (requiring foo)

Reply:
> Is that agreeable?

Sure, it is.

To answer to your `Hint Cut` question, I'm far from being an expert but I would say, if you absolutely need to introduce a `Hint Cut` that will only be useful for the first proof, yes you can work your way with locality and modules or sections, but usually when you introduce a `Hint Cut` it is because this is something that is *always* desirable.